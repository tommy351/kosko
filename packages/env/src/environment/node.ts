import { getRequireExtensions } from "@kosko/require";
import { join } from "node:path";
import { formatPath } from "../paths";
import { toArray } from "@kosko/utils";
import {
  createEnvironment,
  createLoaderReducers,
  createReducerList
} from "./base";
import { Environment, Loader, ReducerExecutor, ReducerList } from "./types";

export function getNodeExtensions(): string[] {
  return getRequireExtensions().map((ext) => ext.substring(1));
}

/**
 * @public
 */
export interface NodeEnvironmentOptions {
  /**
   * Current working directory.
   */
  cwd?: string;
}

export interface InternalNodeEnvironmentOptions extends NodeEnvironmentOptions {
  createReducerExecutor(reducers: ReducerList): ReducerExecutor;
  requireModule(environment: Environment, id: string): unknown;
  mergeValues(data: unknown[]): unknown;
}

export function createNodeEnvironment({
  cwd = process.cwd(),
  createReducerExecutor,
  requireModule,
  mergeValues
}: InternalNodeEnvironmentOptions): Environment {
  function requireAllEnvs(template: string, component?: string) {
    if (!environment.env) return [];

    const envs = toArray(environment.env);

    return envs.map((env) => {
      const path = formatPath(template, {
        environment: env,
        ...(component && { component })
      });

      return requireModule(environment, join(environment.cwd, path));
    });
  }

  const loader: Loader = {
    global: () => requireAllEnvs(environment.paths.global),
    component: (name) => requireAllEnvs(environment.paths.component, name)
  };
  const reducers = createReducerList(createLoaderReducers(loader, mergeValues));
  const { reduce } = createReducerExecutor(reducers);
  const environment = createEnvironment({
    cwd,
    extensions: getNodeExtensions(),
    setReducers: (cb) => reducers.setReducers(cb),
    resetReducers: () => reducers.resetReducers(),
    reduce
  });

  return environment;
}
