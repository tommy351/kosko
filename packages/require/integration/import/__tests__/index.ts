import execa, { NodeOptions } from "execa";
import { dirname, join } from "node:path";

const testDir = dirname(__dirname);

function parseResult(s: string) {
  const data = JSON.parse(s);

  // Node.js 23 adds `module.exports` export on ESM CJS wrapper. Omit it for
  // consistent snapshots.
  // https://github.com/nodejs/node/pull/53848
  delete data["module.exports"];

  return data;
}

describe("Node.js CJS", () => {
  async function execute(options: NodeOptions) {
    return execa.node(join(testDir, "entry-node.cjs"), [], {
      cwd: testDir,
      ...options
    });
  }

  test("should import MJS file", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "file-mjs.mjs" }
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import CJS file", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "file-cjs.cjs" }
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import JSON file", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "file-json.json" }
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import MTS file", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "file-mts.mts" },
      nodeOptions: ["--loader", "ts-node/esm"]
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import JS file in package whose type = commonjs", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "pkg-commonjs/file-js.js" }
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import JS file in package whose type = module", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "pkg-module/file-js.js" }
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import TS file in package whose type = module", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "pkg-module/file-ts.ts" },
      nodeOptions: ["--loader", "ts-node/esm"]
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });
});

describe("Node.js ESM", () => {
  async function execute(options: NodeOptions) {
    return execa.node(join(testDir, "entry-node.mjs"), [], {
      cwd: testDir,
      ...options
    });
  }

  test("should import MJS file", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "file-mjs.mjs" }
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import CJS file", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "file-cjs.cjs" }
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import JSON file", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "file-json.json" }
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import MTS file", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "file-mts.mts" },
      nodeOptions: ["--loader", "ts-node/esm"]
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import JS file in package whose type = commonjs", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "pkg-commonjs/file-js.js" }
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import JS file in package whose type = module", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "pkg-module/file-js.js" }
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });

  test("should import TS file in package whose type = module", async () => {
    const result = await execute({
      env: { IMPORT_PATH: "pkg-module/file-ts.ts" },
      nodeOptions: ["--loader", "ts-node/esm"]
    });

    expect(parseResult(result.stdout)).toMatchSnapshot();
  });
});
