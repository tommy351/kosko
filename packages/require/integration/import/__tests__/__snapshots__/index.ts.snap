// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node.js CJS should import CJS file 1`] = `
{
  "default": {
    "ext": "mjs",
  },
  "ext": "mjs",
}
`;

exports[`Node.js CJS should import JS file in package whose type = commonjs 1`] = `
{
  "default": {
    "ext": "js",
  },
  "ext": "js",
}
`;

exports[`Node.js CJS should import JS file in package whose type = module 1`] = `
{
  "ext": "js",
}
`;

exports[`Node.js CJS should import JSON file 1`] = `
{
  "default": {
    "ext": "json",
  },
}
`;

exports[`Node.js CJS should import MJS file 1`] = `
{
  "ext": "mjs",
}
`;

exports[`Node.js CJS should import MTS file 1`] = `
{
  "ext": "mts",
}
`;

exports[`Node.js CJS should import TS file in package whose type = module 1`] = `
{
  "ext": "ts",
}
`;

exports[`Node.js ESM should import CJS file 1`] = `
{
  "default": {
    "ext": "mjs",
  },
  "ext": "mjs",
}
`;

exports[`Node.js ESM should import JS file in package whose type = commonjs 1`] = `
{
  "default": {
    "ext": "js",
  },
  "ext": "js",
}
`;

exports[`Node.js ESM should import JS file in package whose type = module 1`] = `
{
  "ext": "js",
}
`;

exports[`Node.js ESM should import JSON file 1`] = `
{
  "default": {
    "ext": "json",
  },
}
`;

exports[`Node.js ESM should import MJS file 1`] = `
{
  "ext": "mjs",
}
`;

exports[`Node.js ESM should import MTS file 1`] = `
{
  "ext": "mts",
}
`;

exports[`Node.js ESM should import TS file in package whose type = module 1`] = `
{
  "ext": "ts",
}
`;
